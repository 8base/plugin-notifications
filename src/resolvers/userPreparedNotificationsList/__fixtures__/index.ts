export const TABLE_SCHEMA = {
  id: '5e3836df9564db0007a4a525',
  fields: [
    {
      name: 'id',
      displayName: 'ID',
      fieldType: 'ID',
      isSystem: true,
      relation: null,
    },
    {
      name: 'createdAt',
      displayName: 'Created At',
      fieldType: 'DATE',
      isSystem: true,
      relation: null,
    },
    {
      name: 'updatedAt',
      displayName: 'Updated At',
      fieldType: 'DATE',
      isSystem: true,
      relation: null,
    },
    {
      name: 'createdBy',
      displayName: 'Created By',
      fieldType: 'RELATION',
      isSystem: true,
      relation: {
        refTable: {
          name: 'Users',
          fields: [
            {
              id: '5e381797dcb2b80008734828',
              name: 'id',
              fieldType: 'ID',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734829',
              name: 'createdAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b8000873482c',
              name: 'updatedAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e38179bdcb2b800087349b9',
              name: 'createdBy',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734832',
              name: 'email',
              fieldType: 'TEXT',
              isSystem: true,
            },
            {
              id: '5e38179ddcb2b80008734a10',
              name: 'status',
              fieldType: 'SWITCH',
              isSystem: true,
            },
            {
              id: '5e38179ddcb2b80008734a13',
              name: 'origin',
              fieldType: 'SWITCH',
              isSystem: true,
            },
            {
              id: '5e38179bdcb2b800087349b6',
              name: 'is8base',
              fieldType: 'SWITCH',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734838',
              name: 'firstName',
              fieldType: 'TEXT',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b8000873483b',
              name: 'lastName',
              fieldType: 'TEXT',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b8000873483e',
              name: 'timezone',
              fieldType: 'TEXT',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734943',
              name: 'avatar',
              fieldType: 'FILE',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734956',
              name: 'roles',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e381be2dcb2b800087524ea',
              name: 'sentNotifications',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e3a7ceed8955e00087a635e',
              name: 'clientNotifications',
              fieldType: 'RELATION',
              isSystem: false,
            },
          ],
        },
      },
    },
    {
      name: 'notification',
      displayName: 'Notification',
      fieldType: 'RELATION',
      isSystem: false,
      relation: {
        refTable: {
          name: 'Notifications',
          fields: [
            {
              id: '5e381adf8caf0200089283ba',
              name: 'id',
              fieldType: 'ID',
              isSystem: true,
            },
            {
              id: '5e381adf8caf0200089283bb',
              name: 'createdAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381adf8caf0200089283be',
              name: 'updatedAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381adf8caf0200089283c4',
              name: 'createdBy',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e381be2dcb2b800087524e7',
              name: 'actor',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e381b488caf02000892ada2',
              name: 'template',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e38375b9564db0007a4ee0d',
              name: 'entity',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e3a7d3dd8955e00087ab73b',
              name: 'clientNotifications',
              fieldType: 'RELATION',
              isSystem: false,
            },
          ],
        },
      },
    },
    {
      name: 'post',
      displayName: 'Post',
      fieldType: 'RELATION',
      isSystem: false,
      relation: {
        refTable: {
          name: 'Posts',
          fields: [
            {
              id: '5e381a36d1978f00089e37f0',
              name: 'id',
              fieldType: 'ID',
              isSystem: true,
            },
            {
              id: '5e381a36d1978f00089e37f1',
              name: 'createdAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381a36d1978f00089e37f4',
              name: 'updatedAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381a36d1978f00089e37fa',
              name: 'createdBy',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e381a4e8caf020008922bf1',
              name: 'title',
              fieldType: 'TEXT',
              isSystem: false,
            },
            {
              id: '5e381a54dcb2b80008744cf3',
              name: 'message',
              fieldType: 'TEXT',
              isSystem: false,
            },
            {
              id: '5e3837209564db0007a4cbca',
              name: 'notificationEntities',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e3839ee3e61f90007d9ce68',
              name: 'hidden',
              fieldType: 'SWITCH',
              isSystem: false,
            },
            {
              id: '5e396a82be370200087caef2',
              name: 'comments',
              fieldType: 'RELATION',
              isSystem: false,
            },
          ],
        },
      },
    },
    {
      name: 'comment',
      displayName: 'Comment',
      fieldType: 'RELATION',
      isSystem: false,
      relation: {
        refTable: {
          name: 'Comments',
          fields: [
            {
              id: '5e396a61be370200087c8745',
              name: 'id',
              fieldType: 'ID',
              isSystem: true,
            },
            {
              id: '5e396a61be370200087c8746',
              name: 'createdAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e396a61be370200087c8749',
              name: 'updatedAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e396a61be370200087c874f',
              name: 'createdBy',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e396a6f97c4ef0007ebcf0f',
              name: 'message',
              fieldType: 'TEXT',
              isSystem: false,
            },
            {
              id: '5e396a82be370200087caeef',
              name: 'post',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e396aea1f790b00082e6a7a',
              name: 'notificationEntities',
              fieldType: 'RELATION',
              isSystem: false,
            },
          ],
        },
      },
    },
  ],
};

export const NOTIFICATION_ENTITIES_SCHEMA_RESPONSE = {
  system: {
    table: TABLE_SCHEMA,
  },
};

export const NOTIFICATIONS_LIST_RESPONSE = {
  userNotificationsList: {
    items: [
      {
        notification: {
          entity: {
            post: {
              id: 'postId',
              title: 'post title',
              message: 'post message',
            },
          },
          template: {
            entityType: 'Posts',
            title: 'New post "{{ post.title }}"',
            message: '{{ actor.firstName }} {{ actor.lastName }} created a new post "{{ post.title }}"',
          },
          actor: {
            id: 'actorId',
            email: 'zouxuoz@gmail.com',
            firstName: 'Vladimir',
            lastName: 'Osipov',
          },
        },
        id: 'notficationId',
        readed: false,
      },
    ],
  },
};
