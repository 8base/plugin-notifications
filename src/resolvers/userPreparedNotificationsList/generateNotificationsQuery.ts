import * as R from 'ramda';
import { print, SelectionSetNode, FieldNode } from 'graphql';

import { USER_NOTIFICATIONS_LIST_QUERY } from './queries';

export type TableFieldSchema = {
  name: string;
  relation?: {
    refTable: {
      id: string;
      name: string;
    };
  };
  fieldType: string;
  fieldTypeAttributes?: {
    format: string;
    innerFields: Array<{
      name: string;
      fieldType: string;
    }>;
  } | null;
};

export type TableSchema = object;

const getTableByName = (tablesSchema: TableSchema[], tableName: string): TableSchema | undefined =>
  R.find<TableSchema>(R.propEq('name', tableName), tablesSchema);

const getTableFieldsByName = (tablesSchema: TableSchema[], tableName: string): TableFieldSchema[] => {
  const table = getTableByName(tablesSchema, tableName);

  if (!table) {
    return [];
  }

  return R.propOr([], 'fields', table);
};

const isRelationField = (tableField: TableFieldSchema): boolean => R.propEq('fieldType', 'RELATION', tableField);
const isFileField = (tableField: TableFieldSchema): boolean => R.propEq('fieldType', 'FILE', tableField);
const isSmartField = (tableField: TableFieldSchema): boolean => R.propEq('fieldType', 'SMART', tableField);

const generateTableFieldNode = (
  tableField: TableFieldSchema,
  tablesSchema: TableSchema[],
  filters?: Array<(field: TableFieldSchema) => boolean>,
  depth = 0,
): FieldNode => {
  let field: FieldNode = {
    kind: 'Field',
    name: {
      kind: 'Name',
      value: tableField.name,
    },
    arguments: [],
    directives: [],
  };

  if (isRelationField(tableField) && depth <= 3) {
    const refTableName: string | undefined = R.path(['relation', 'refTable', 'name'], tableField);

    if (refTableName) {
      field = R.assoc(
        'selectionSet',
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        generateTableFieldsSelectionSet(tablesSchema, refTableName, filters, ++depth),
        field,
      );
    }
  } else if (isFileField(tableField) && depth <= 3) {
    const refTableName = 'Files';

    if (refTableName) {
      field = R.assoc(
        'selectionSet',
        {
          kind: 'SelectionSet',
          selections: ['id', 'uploadUrl', 'downloadUrl', 'shareUrl', 'filename'].map(name => ({
            kind: 'Field',
            name: {
              kind: 'Name',
              value: name,
            },
            arguments: [],
            directives: [],
          })),
        },
        field,
      );
    }
  } else if (isSmartField(tableField) && depth <= 3) {
    const innerFields = R.pathOr([], ['fieldTypeAttributes', 'innerFields'], tableField);

    field = R.assoc(
      'selectionSet',
      {
        kind: 'SelectionSet',
        selections: innerFields.map(({ name }) => ({
          kind: 'Field',
          name: {
            kind: 'Name',
            value: name,
          },
          arguments: [],
          directives: [],
        })),
      },
      field,
    );
  }

  return field;
};

const generateTableFieldsSelectionSet = (
  tablesSchema: TableSchema[],
  tableName: string,
  filters?: Array<(field: TableFieldSchema) => boolean>,
  depth = 0,
): SelectionSetNode => {
  const tableFields = getTableFieldsByName(tablesSchema, tableName);

  const selections = R.pipe<TableFieldSchema[], TableFieldSchema[], FieldNode[]>(
    R.reject(filters && filters[depth] ? filters[depth] : R.F),
    R.map((field: TableFieldSchema) => generateTableFieldNode(field, tablesSchema, filters, depth)),
  )(tableFields);

  return {
    kind: 'SelectionSet',
    selections,
  };
};

export const generateNotificationsQuery = (tablesSchema: TableSchema[]) => {
  const selectionSet: SelectionSetNode = generateTableFieldsSelectionSet(tablesSchema, 'NotificationEntities', [
    R.anyPass([
      R.propEq('isSystem', true),
      R.propEq('isList', true),
      R.propEq('name', 'notification'),
      R.complement(R.propEq('fieldType', 'RELATION')),
    ]),
    R.anyPass([R.propEq('isList', true)]),
    R.anyPass([R.propEq('isList', true), R.propEq('fieldType', 'RELATION')]),
  ]);

  const queryLens = R.lensPath([
    'definitions',
    0,
    'selectionSet',
    'selections',
    0,
    'selectionSet',
    'selections',
    0,
    'selectionSet',
    'selections',
    0,
    'selectionSet',
    'selections',
    2,
  ]);

  return print(
    R.over(
      queryLens,
      R.append({
        kind: 'Field',
        name: {
          kind: 'Name',
          value: 'entity',
        },
        arguments: [],
        directives: [],
        selectionSet,
      }),
      USER_NOTIFICATIONS_LIST_QUERY,
    ),
  );
};
