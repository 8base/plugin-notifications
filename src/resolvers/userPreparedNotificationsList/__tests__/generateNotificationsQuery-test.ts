import { generateNotificationsQuery } from '../generateNotificationsQuery';
import { TABLE_SCHEMA } from '../__fixtures__';

it('Should returns user notifications.', async () => {
  expect(generateNotificationsQuery(TABLE_SCHEMA)).toMatchSnapshot();
});
