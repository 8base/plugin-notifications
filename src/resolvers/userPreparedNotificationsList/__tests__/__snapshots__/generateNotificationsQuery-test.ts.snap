// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should returns user notifications. 1`] = `
"query UserNotifications {
  userNotificationsList {
    items {
      notification {
        template {
          entityType
          title
          message
        }
        actor {
          id
          firstName
          lastName
          email
        }
        entity {
          post {
            id
            createdAt
            updatedAt
            createdBy {
              id
              createdAt
              updatedAt
              email
              status
              origin
              is8base
              firstName
              lastName
              timezone
              avatar {
                id
                uploadUrl
                downloadUrl
                shareUrl
                filename
              }
            }
            title
            message
            hidden
          }
          comment {
            id
            createdAt
            updatedAt
            createdBy {
              id
              createdAt
              updatedAt
              email
              status
              origin
              is8base
              firstName
              lastName
              timezone
              avatar {
                id
                uploadUrl
                downloadUrl
                shareUrl
                filename
              }
            }
            message
            post {
              id
              createdAt
              updatedAt
              title
              message
              hidden
            }
          }
        }
      }
      createdAt
      id
      read
    }
    count
  }
}
"
`;
